wdl-lint/src/rules/command_shellcheck.rs

```diff
@@ -0,0 +1,357 @@
+    pub end_line: usize,
+    /// column comment starts on
+    pub column: usize,
+    /// column comment ends on
+    #[serde(rename = "endColumn")]
+    pub end_column: usize,
+    /// severity of the comment
+    pub level: String,
+    /// shellcheck error code
+    pub code: usize,
+    /// message associated with the comment
+    pub message: String,
+}
+
+/// Run shellcheck
+///
+/// writes command text to stdin of shellcheck process
+/// and returns parsed ShellCheckComments
+fn run_shellcheck(command: &str) -> Result<Vec<ShellCheckComment>> {
+    let mut sc_proc = process::Command::new(SHELLCHECK_BIN)
+        .args([
+            "-s",
+            "bash",
+            "-f",
+            "json",
+            "-e",
+            &SHELLCHECK_SUPPRESS.join(","),
+            "-",
+        ])
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()
+        .context("error calling shellcheck")?;
+    {
+        let mut proc_stdin = sc_proc
+            .stdin
+            .take()
+            .context("error obtaining stdin of shellcheck")?;
+        proc_stdin.write_all(command.as_bytes())?;
+    }
+
+    let output = sc_proc
+        .wait_with_output()
+        .context("error calling shellcheck")?;
+
+    // shellcheck returns exit code 1 if
+    // any checked files result in comments
+    // so cannot check with status.success()
+    match output.status.code() {
+        Some(0) | Some(1) => serde_json::from_slice::<Vec<ShellCheckComment>>(&output.stdout)
+            .context("failed to parse shellcheck output"),
+        Some(code) => bail!("error calling shellcheck: exit code {}", code),
+        None => bail!("error calling check: process interrupted"),
+    }
+}
+
+/// Runs ShellCheck on a command section and reports violations
+#[derive(Default, Debug, Clone, Copy)]
+pub struct CommandSectionShellCheck;
+
+impl Rule for CommandSectionShellCheck {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that command blocks are free of ShellCheck violations."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "ShellCheck is a static analysis tool and linter for sh / bash. \
+        The lints provided by ShellCheck help prevent common errors and \
+        pitfalls in your scripts. Following its recommendations will increase \
+        the robustness of your command blocks."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Correctness, Tag::Style, Tag::Portability])
+    }
+
+    fn exceptable_nodes(&self) -> Option<&'static [SyntaxKind]> {
+        Some(&[
+            SyntaxKind::VersionStatementNode,
+            SyntaxKind::TaskDefinitionNode,
+            SyntaxKind::CommandSectionNode,
+        ])
+    }
+}
+
+/// Convert a WDL placeholder to a bash variable
+fn to_bash_var(placeholder: &Placeholder) -> String {
+    let placeholder_text = placeholder.syntax().text().to_string();
```

Comment:
`text().to_string()` can be expensive to call for complex expressions (it's a recursive traversal of the CST if I'm not mistaken). We might want to rethink this part. Maybe generate a random sequence of ASCII?

Reply:
Agree.